### PR TITLE
upgrade doc

### PR DESCRIPTION
**Pull request by shicholas**
I recently upgraded my ``.8`` ams app to the latest and found the following helpful. Hope someone can find this useful too.

I also feel I'm missing something when it comes to creating an ArraySerializer Adapter and would appreciate some help.

### CODE REVIEW

**Comment by beauby** [comment]
Thanks @shicholas, that's really helpful! It might even make sense to provide a legacy `json_08` adapter in 0.10 to make the migration even easier. I know this is something that has been talked about, but I don't think anyone has taken a crack at it as of yet.
We could discuss about the `ArraySerializer` on our [Slack](http://amserializers.slack.com/).

**Comment by shicholas** [comment]
do I need to an invite to join slack?

**Comment by joaomdmoura** [comment]
@shicholas: https://amserializers.herokuapp.com/

File: docs/howto/upgrade_from_previous_versions.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+### Creating a Base and Collection Adapter
+
+You can upgrade to ams ``.10`` by creating a Base and Collection Adapter to
+emulate the response that your application currently serves.
+
+###### Base Adapter
+
+A base adapter takes a ``ActiveModel::Serializer`` instance, and creates a hash
+used for serialization in its ``serializeable_hash`` method. A base adapter
```
Comment (bf4) [inline docs/howto/upgrade_from_previous_versions.md]: typo `serializable`

File: docs/howto/upgrade_from_previous_versions.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
+# How to upgrade from previous versions
```
Comment (bf4) [inline docs/howto/upgrade_from_previous_versions.md]: I think `UPGRADE.md` or `UPGRADING.md` yes caps might be a better name.

Also, would be good to add to the gemspec

`spec.upgrade = 'See docs/howto/UPGRADE.md for upgrade instructions` (syntax might not be quite right)

**Comment by bf4** [comment]
This would be fantastic as code with a PR and tests. If we can version APIs, why not adapters?

File: docs/howto/upgrade_from_previous_versions.md (hunk 2 above)
Comment (shicholas) [inline docs/howto/upgrade_from_previous_versions.md]: > spec.upgrade = 'See docs/howto/UPGRADE.md for upgrade instructions (syntax might not be quite right)

AFAIK there is nothing for docs aside from [adding to RDoc files](http://guides.rubygems.org/specification-reference/#extra_rdoc_files). None of the other doc files are in the ``.gemspec``, should I add this as a start?

**Comment by shicholas** [comment]
I don't know about version ``.9``, but maybe for ``.8`` we could have these adapters:

* Version8::BaseAdapter (with a root element)
* Version8::BaseAdapterWithoutRoot
* Version8::CollectionAdapter
* Version8::CollectionAdapterWithoutRoot

I feel most apps I've worked with use version ``.8`` with the root adapter, and since ``root`` shouldn't be defined per serializer anymore, I chose these four adapters.

File: docs/howto/UPGRADING.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+###### Base Adapter
+
+A base adapter takes a ``ActiveModel::Serializer`` or
+``ActiveModel::Serializer::ArraySerializer`` instance, and creates a hash
+used for serialization in its ``serializable_hash`` method. A base adapter
+could look like this:
+
+```ruby
+class MyApp::BaseAdapter < ActiveModel::Serializer::Adapter::Base
```
Comment (bf4) [inline docs/howto/UPGRADING.md]: typo: `;:`.  why not try adding this code with tests? 

I'm thinking

```ruby
# lib/active_model_serializers/v08/serializer.rb
# lib/active_model_serializers/v08/collection_serializer.rb
# lib/active_model_serializers/v08/adapter.rb
module ActiveModelSerializers::V08
  # Serializer might not need anything, but good to have defined
  Serializer = ActiveModel::Serializer
  CollectionSerializer = ActiveModel::Serializer::CollectionSerializer # well, it's Array now, but hopefully we'll get that renaming pr soon
  class Adapter < ActiveModel::Serializer::Adapter::Base
    # whatever methods you need
  end
end
```

And test like the other serializers and adapters

```ruby
adapter = :'v08/adapter'
serializer = :'v08/array_serializer'
each_serializer = :'v08/serializer'
collection_resource = Post.all
expected_response = [{ something }]
assert ActiveModel::SerializableResource.new(collection_resource, serializer: serializer, each_serializer: each_serializer, adapter: adapter).as_json, expected_response

# and

adapter = :'v08/adapter'
serializer = :'v08/serializer'
resource = Post.first
expected_response = { something }
assert ActiveModel::SerializableResource.new(collection_resource, serializer: serializer, adapter: adapter).as_json, expected_response
```

Comment (bf4) [inline docs/howto/UPGRADING.md]: Though after clicking comment, I realized the options definitions are wrong, should be:

```ruby
adapter = ActiveModelSerializers::V08::Adapter
serializer = ActiveModelSerializers::V08::ArraySerializer'
each_serializer = ActiveModelSerializers::V08::Serializer
```

Comment (bf4) [inline docs/howto/UPGRADING.md]: @rails-api/ams thoughts on naming, folder etc?

Comment (NullVoxPopuli) [inline docs/howto/UPGRADING.md]: anything specific? looks ok to me :-\ (except for the crazy long asserts)

Comment (shicholas) [inline docs/howto/UPGRADING.md]: the API for the serializer itself changes (namely the ``root`` option). How should I address that?

Comment (shicholas) [inline docs/howto/UPGRADING.md]: @bf4: 

> typo: ;:. why not try adding this code with tests? 

sorry I must be blind, I don't see the typo.

Comment (bf4) [inline docs/howto/UPGRADING.md]: Code blind 
Semicolon in place of colon

Way i wrote was ambiguous 

B mobile phone

**Comment by shicholas** [comment]
closing, I think a V8 adapter is a good idea (though still would like to know how to deal with the issue of the ``root`` option) and I feel the docs that will result from https://github.com/rails-api/active_model_serializers/issues/1317 will trump this PR. 

Also, I feel with the code somewhat in flux, e.g. the namespacing changes I have been reading, maybe the v8 adapter should be written after the api stabilizes?

**Pull request closed without merging**